### PR TITLE
Test edge cases

### PR DESCRIPTION
**Pull request by erikkemperman**
(Don't merge, broken on win32).

I thought I might clean up and contribute some of my stumbling around the edges of the node-glob update and corresponding ignore option shuffle.

This is the syntax, at least if I've now understood previous discussion properly, that users would need for several edge cases. Agree?

Unfortunately it is slightly unpractical for me to debug why this breaks on windows, exactly, but I'd guess that somewhere it treats our escape character as path separator. If so, I think that ought to be addressed? 

Travis [passes](https://travis-ci.org/erikkemperman/glob-stream/builds/161033064),
Appveyor [doesn't](https://ci.appveyor.com/project/erikkemperman/glob-stream/build/1.0.12).

### CODE REVIEW

**Comment by phated** [comment]
Edge cases for the edge-case-god.  Let's wait on these until some of the other issues are solved.  I might be removing `is-absolute-glob` once https://github.com/isaacs/node-glob/pull/293 is resolved.

**Comment by erikkemperman** [comment]
Fair enough, that sounds good.

**Comment by phated** [comment]
@erikkemperman could you rebase this and update to the new style of tests in master?  Thanks!

**Comment by erikkemperman** [comment]
@phated Yup, will do just as soon as I have some time!

**Comment by erikkemperman** [comment]
@phated Rebased and restyled... Still passing on Travis, but still failing on AppVeyor. Can't dig into that further though, for lack of a windows box.

**Comment by phated** [comment]
Thanks!  cc'ing @jonschlinkert and @doowb about this, maybe they can help since we are keeping `to-absolute-glob` as a dependency.

**Comment by doowb** [comment]
I'll be able to take a look and debug on a windows machine over the next few days.

**Comment by phated** [comment]
@doowb perfect, that's the timeframe I'm looking at for the next glob-stream major.

File: test/main.js [hunk 1]
```diff
@@ -666,6 +666,150 @@ describe('glob-stream', function() {
...
+      path: cwd + '/case',
+    };
+
+    function assert(pathObjs) {
+      expect(pathObjs.length).toEqual(1);
+      expect(pathObjs[0]).toEqual(expected);
+    }
+
+    pipe([
+      globStream(['*', '!!(case)'], { cwd: cwd }),
```
Comment (jonschlinkert) [inline test/main.js]: a double `!`  would be recognized as an inclusive/positive matcher by minimatch. From the [minimatch readme](https://github.com/isaacs/minimatch#comparisons-to-other-fnmatchglob-implementations)

> Multiple `!` characters at the start of a pattern will negate the pattern multiple times.

Comment (erikkemperman) [inline test/main.js]: From earlier discussion I took this syntax to mean "ignore everything matching this negative extglob", which is what it tests

 All of these pass on Travis, but not on Appveyor. I'm thinking slashes get parsed as separators, but am not equipped to debug on windows...

Comment (jonschlinkert) [inline test/main.js]: I just finished running the tests on Windows 8, Windows 10 and mac with node.js 6 and the current version, and everything is passing 100%. I also logged out the paths created for the failing unit tests, and everything looks correct to me. 

I'm waiting for nvm to finish installing on my windows box so I can test other versions. I'll update here when I get those results. 

windows-8

Comment (phated) [inline test/main.js]: I don't see the added tests in that output (?)

Comment (erikkemperman) [inline test/main.js]: Thanks, I just noticed the same thing -- but accidentally started a review or something. Tired fingers on tiny mobile screen, sorry about that.

Comment (jonschlinkert) [inline test/main.js]: yeah @doowb just said that. I had a few different screens and pasted the wrong one for master. I did test Erik's branch too. I'm still trying to get nvm to work on windows (the go version). I'll update once I get that working.

Comment (jonschlinkert) [inline test/main.js]: also, I'll need to confirm this but we might need to convert escaped patterns to double-backslashes on windows. Since a backslash will be regarded as a file path I believe.

Comment (erikkemperman) [inline test/main.js]: Yeah that was my original guess, slashes parsed as separators on windows. Sorry I'm not able to track this down myself, sounds like I'm claiming a lot of your time...

Comment (jonschlinkert) [inline test/main.js]: oh no worries, I don't mind. This windows path stuff can be difficult to debug, I love being a part of a community where other devs can help me figure this stuff out. I'm happy to help where I can

Comment (jonschlinkert) [inline test/main.js]: quick update. I tried for a couple of hours to get [nvm-windows](https://github.com/coreybutler/nvm-windows) working on my windows box with no luck. I don't think it was a windows-nvm issue, but I'm going to try again tomorrow and let you know once I get these tests running on different versions. Unless someone has a better suggestion for testing node.js versions on windows.

Comment (phated) [inline test/main.js]: @jonschlinkert I'm not sure this is a node version issue, it seems to be failing on every version.

**Comment by phated** [comment]
@doowb any luck debugging? I'm getting super close to cutting the 6.0 release of glob-stream.

File: test/main.js (hunk 1 above)
Comment (phated) [inline test/main.js]: Side note: this test seems to be passing on Appveyor, if I'm reading the results correctly.

File: test/main.js [hunk 2]
```diff
@@ -666,6 +666,150 @@ describe('glob-stream', function() {
...
+      expect(pathObjs[0]).toEqual(expected);
+    }
+
+    pipe([
+      globStream('\\!case', { cwd: cwd }),
+      concat(assert),
+    ], done);
+  });
+
+  it('should handle pattern starting with exclam and paren', function(done) {
```
Comment (phated) [inline test/main.js]: @jonschlinkert @doowb this seems to be one of the failing tests

File: test/main.js [hunk 3]
```diff
@@ -666,6 +666,150 @@ describe('glob-stream', function() {
...
+      expect(pathObjs.length).toEqual(0);
+    }
+
+    pipe([
+      globStream(['*', '!\\!case'], { cwd: cwd }),
+      concat(assert),
+    ], done);
+  });
+
+  it('should ignore pattern starting with exclam and paren', function(done) {
```
Comment (phated) [inline test/main.js]: @jonschlinkert @doowb this seems to be the other failing test

**Comment by erikkemperman** [comment]
@phated Looks like you renamed `test/main.js` to `test/index.js`? Should I rebase, or better to await review/debug?

**Comment by phated** [comment]
@erikkemperman yeah, I reorganized the project. Feel free to rebase if you have time now (otherwise I was going to do it in a branch).  I believe `master` is complete on my end.

File: test/main.js (hunk 3 above)
Comment (jonschlinkert) [inline test/main.js]: yeah they're both essentially the same issue. I believe node-glob or minimatch strips escaping and/or normalize all `\\` to `/`, so no matter how we do the pattern there it won't work. But I can't get nvm working on windows to confirm.

**Comment by phated** [comment]
@erikkemperman I wonder if the 2 failing tests are related to https://www.npmjs.com/package/glob-parent#windows - maybe @es128 would know

File: test/main.js (hunk 3 above)
Comment (jonschlinkert) [inline test/main.js]: and if it's not that, then it's probably how the tests are designed. If node-glob doesn't normalize slashes then wouldn't we need to do double backslashes on windows? Otherwise the current escaping would be recognized as a path separator.

Comment (phated) [inline test/main.js]: I was thinking it might be https://www.npmjs.com/package/glob-parent#windows (mentioned below in the thread)

**Comment by jonschlinkert** [comment]
> I wonder if the 2 failing tests are related to https://www.npmjs.com/package/glob-parent#windows - maybe @es128 would know

That's not a glob-parent issue, [it's a node-glob issue](https://github.com/isaacs/node-glob#windows)

**Comment by phated** [comment]
@jonschlinkert not quite - remember https://github.com/es128/glob-parent/pull/13

**Comment by phated** [comment]
@erikkemperman @jonschlinkert @doowb I'm trying the glob-parent thing at https://github.com/gulpjs/glob-stream/commit/6715d6b7801c626512f13677a03ccb60377672cb to see if anything changes on windows (yay appveyor)

**Comment by jonschlinkert** [comment]
also, fwiw is exactly why I've stated a bunch of times that we should try to do anything at all to work around these backslash issues. I was concerned that it would make it harder for users to figure out where the issue was being caused. But it's making it harder for us to figure it out too.

**Comment by phated** [comment]
I believe I agreed and wanted to tell people to use `/` only (as node-glob does) but that wasn't desirable from @es128's perspective.

**Comment by phated** [comment]
My test didn't work 😛

**Comment by jonschlinkert** [comment]
> I believe I agreed and wanted to tell people to use / only (as node-glob does) but that wasn't desirable from @es128's perspective.

true, good point. I'm not sure what we can do to fix this (other than using a different glob lib)

**Comment by doowb** [comment]
I just logged out the paths in different places in windows and the glob pattern that's being based to `node-glob` is `C:/Users/brian/work/gulpjs/glob-stream/test/fixtures/edge/\!(case)` which returns:

```
C:/Users/brian/work/gulpjs/glob-stream/test/fixtures/edge/!(case)
C:/Users/brian/work/gulpjs/glob-stream/test/fixtures/edge/!case
C:/Users/brian/work/gulpjs/glob-stream/test/fixtures/edge/(case)
```

I'm going to see if "doubling up" the escape characters on windows does anything (like @jonschlinkert  mentioned above), but this is something that `node-glob` isn't handling correctly for windows.

**Comment by doowb** [comment]
Same results when the pattern looks like this: `C:/Users/brian/work/gulpjs/glob-stream/test/fixtures/edge/\\!(case)`

**Comment by es128** [comment]
I'm fine with ripping out the backslash auto-fix thing with a major version bump if it's causing other problems. 

But is that what's happening here?

**Comment by phated** [comment]
@es128 I don't think it's a glob-parent problem because I attempted the `./` prefix thing.

**Comment by jonschlinkert** [comment]
> But is that what's happening here?

I don't think so. I think it's node-glob.

**Comment by es128** [comment]
Right, I don't think it applies. But I don't quite understand what the test is trying to test. A directory literally named `!(case)`? Perhaps `\\!\\(case\\)`

**Comment by phated** [comment]
@doowb interesting!!!! What are the results for the other test because that is getting the inverse and might lead us to the solution

**Comment by doowb** [comment]
I updated the comments above with the correct amount of `\` in the glob patterns. One second on the other tests...

**Comment by erikkemperman** [comment]
FWIW I've rebased this branch onto tip of master at https://github.com/erikkemperman/glob-stream/tree/test-edge-cases-rb
Same test results, I've just appended the edge case tests to the new `test/index.js`.

About the slash/separator ambiguity: remember these tests just reflect my understanding of the syntax emerging from earlier discussion -- so maybe let's take a step back and double check if I am actually testing the desired behaviour? Although they pass on Travis, so there's that :-)

**Comment by phated** [comment]
@erikkemperman I don't know that we've ever talked about parens at the end of a path, I've only ever considered them in the directory position

**Comment by phated** [comment]
@erikkemperman please rebase into this branch. I've also split your tests into a separate `describe` named "edge cases"

**Comment by phated** [comment]
Thinking more about this, we don't even know if `node-glob` handles globs that contain `( )` correctly and we don't test that at all (nor should we).  The only thing we test is that `to-absolute-glob` and `glob-parent` don't see `( )` as glob characters in the directory portion of the glob and return us the end of the pattern.  I honestly think we can do away with those tests and maybe all of these edge cases because they are only testing `node-glob` behavior. @erikkemperman @jonschlinkert @doowb @es128 thoughts?

**Comment by phated** [comment]
[`minimatch` does something with `(`](https://github.com/isaacs/minimatch/blob/master/minimatch.js#L375) but I'm guessing it's wrong on Windows or something like that

**Comment by doowb** [comment]
results of one that's passing:

```
ourGlob: 'C:/Users/brian/work/gulpjs/glob-stream/test/fixtures/edge/\\!case'
matched: 'C:/Users/brian/work/gulpjs/glob-stream/test/fixtures/edge/!case'
```

**Comment by es128** [comment]
Testing things that are the responsibility of your dependencies makes sense to a point so you can easily swap them out in the future. But going deep on contrived edge cases, unless these are cases that have come from real users, may not.

**Comment by phated** [comment]
Removing the 6.0 milestone on this because I think it's going to end up being closed without merging.  Let's continue to discuss if anyone has more thoughts on this.

**Comment by erikkemperman** [comment]
Rebased but coming back to this thread it might no longer be needed.

I'm not even sure these are valid file names on Windows, and if so if git handles them correctly. Does checking out this branch on Windows even yield all 4 files in `/test/fixtures/edge`?

Anyway, fair enough if I'm actually testing the wrong edges :-) Feel free to just close this, or I guess we could skip if on Windows.

**Comment by phated** [comment]
An aside: Maybe we should be testing these cases as directory segments of a glob instead of the glob itself?

**Comment by erikkemperman** [comment]
Well, what I intended to make explicit here is my mis?)understanding of how parens and/or exclams would be handled at the beginning of a pattern, because the exclam means something in `glob-stream` that it no longer does in `node-glob` (ignore option) after recent upgrade.

**Comment by erikkemperman** [comment]
So re: your suggestion of edge case in directory segments, I think the significance of _beginning_ in my previous post is that I'd expect these would apply the same way for `!(case)/god` et. al. if there were four directories with one file each.

**Comment by phated** [comment]
@erikkemperman I understand that `!` means we turn a pattern into an ignore but why test parens too?  That doesn't seem necessary if you are trying to test cases negative glob patterns for us, right?

I don't know if ending with vs having the `!( )` as a directory path have different results since we resolve to an absolute path before passing to node-glob, but that probably doesn't need to be tested either.

**Comment by erikkemperman** [comment]
@phated IIRC the discussion earlier turned to the potential confusion between a pattern to go into the ignore option of `node-glob` (with exclam stripped) if a leading exclam preceeds a paren, versus expecting it to go as a regular glob argument unchanged, meaning a negative extglob.

And the (slightly idiotic *) possibility of wanting to match literal exclams and parens in actual paths. Hence the combinations.

But I may very well have things backwards, yay GMT+1.

(*) I've had the misfortune of having to develop for folks who thought such things were perfectly reasonable names for their files.

**Comment by jonschlinkert** [comment]
> I honestly think we can do away with those test

totally agree, @es128 said it best

**Comment by erikkemperman** [comment]
Sorry for trailing off there, sleeeep.

So just to reiterate, I think that these tests do belong here, if anywhere, because they make explicit an API difference with latest `node-glob`. Having said that, they are absolutely far-fetched -- and I am personally never going to run into something like this failing on Windows.

I'll leave it up to @phated to close this PR unmerged, or maybe just skip them on Windows, and apologize to everyone for having wasted your time on this!

**Comment by phated** [comment]
@erikkemperman before we test that, you should create the pattern + ignore tests in node-glob to make sure they are working on Windows.  If those tests work and ours don't, __only then__ are we transforming `!` + `( )` improperly.

**Comment by erikkemperman** [comment]
@phated That'd be interesting, sure! I'll probably find some time tomorrow. Will have to cheat slightly though, because node-glob has an unrelated failing test on Appveyor -- I'll base this on an open [PR](https://github.com/isaacs/node-glob/pull/303).

**Comment by jonschlinkert** [comment]
> apologize to everyone for having wasted your time on this!

fwiw I don't feel that way at all. It is what it is. Windows paths have always been difficult to work with, but we have to deal with it and just keep looking for ways to minimize the potential for regressions. I can't speak for anyone else, but I don't know everything there is to know about ~this~ anything, These discussions help me learn.

**Comment by jonschlinkert** [comment]
> I think that these tests do belong here, if anywhere, because they make explicit an API difference with latest node-glob.

Forgot to mention, I tested the failing patterns directly with node-glob and got the same result that the tests are getting. Both on mac and windows. I'm not sure if you're getting something different with node-glob, but if so I'm at a loss.

**Comment by erikkemperman** [comment]
So, I spent some time with `node-glob` and `minimatch` and you were right -- calling into those directly, I see the same 2 cases fail on Windows in much the same way as via `glob-stream`, while they are working on Linux/Mac..

And I've tracked down that it is indeed because it treats escape chars (backslash) as path separators:

https://github.com/isaacs/minimatch/blob/master/minimatch.js#L123-L125

Several issue reports and proposals to resolve this have been submitted:

https://github.com/isaacs/minimatch/issues/64
https://github.com/isaacs/minimatch/issues/78
https://github.com/isaacs/minimatch/pull/103
(probably more)

But these would not appear to be going anywhere, maintainer is probably too busy with other stuff. I don't think there's much we can do here except hoping Windows users won't want to match literal exclams and parens.

**Comment by phated** [comment]
@erikkemperman @jonschlinkert thanks for testing those cases in node-glob.  I think I'm going to close this because it's really out of our control and there isn't a good way for us to test that we are grouping things correctly when glob isn't working correctly.

**Comment by erikkemperman** [comment]
Fair enough! Too bad Isaac doesn't get around to addressing issues very often, or delegate. Escaping is currently just not possible in minimatch on windows!

**Comment by phated** [comment]
@erikkemperman That's exceptionally frightening!  I agree with you and wish he'd add some contributors to the projects.
